include/picongpu/plugins/Emittance.hpp

```diff
@@ -0,0 +1,1013 @@
+            //! short description of the plugin
+            std::string const description = "calculate the slice emittance of a species";
+            //! prefix used for command line arguments
+            std::string const prefix = ParticlesType::FrameType::getName( ) + std::string( "_emittance" );
+        };
+
+        //! must be implemented by the user
+        static std::shared_ptr< plugins::multi::IHelp > getHelp()
+        {
+            return std::shared_ptr< plugins::multi::IHelp >( new Help{ } );
+        }
+
+        CalcEmittance(
+            std::shared_ptr< plugins::multi::IHelp > & help,
+            size_t const id,
+            MappingDesc* cellDescription
+        ) :
+            m_help( std::static_pointer_cast< Help >(help) ),
+            m_id( id ),
+            m_cellDescription( cellDescription )
+        {
+            filename = m_help->getOptionPrefix() + "_" + m_help->filter.get( m_id ) + ".dat";
+
+            // reduce in same x-z plane
+            constexpr uint32_t r_element = 1u; // y-direction
+
+            /* reduce-add particle properties from other GPUs in same plane: range [r;r+dr]
+             * to "lowest" node in range
+             * e.g.: slice emittance in space y: reduce-add all nodes with same y range in
+             *                         spatial x and z direction to node with
+             *                         lowest x and z position ("corner") and same x range
+             */
+            pmacc::GridController<simDim>& gc = pmacc::Environment<simDim>::get().GridController();
+            pmacc::math::Size_t<simDim> gpuDim = gc.getGpuNodes();
+            pmacc::math::Int<simDim> gpuPos = gc.getPosition();
+
+            /* my plane means: the r_element I am calculating should be 1GPU in width */
+            pmacc::math::Size_t<simDim> sizeTransversalPlane(gpuDim);
+            sizeTransversalPlane[r_element] = 1;
+
+            for( int planePos = 0; planePos <= (int)gpuDim[r_element]; ++planePos )
+            {
+                /* my plane means: the offset for the transversal plane to my r_element
+                 * should be zero
+                 */
+                pmacc::math::Int<simDim> longOffset(pmacc::math::Int<simDim>::create(0));
+                longOffset[r_element] = planePos;
+
+                zone::SphericZone<simDim> zoneTransversalPlane( sizeTransversalPlane, longOffset );
+
+                /* Am I the lowest GPU in my plane? */
+                bool isGroupRoot = false;
+                bool isInGroup   = ( gpuPos[r_element] == planePos );
+                if( isInGroup )
+                {
+                    pmacc::math::Int<simDim> inPlaneGPU(gpuPos);
+                    inPlaneGPU[r_element] = 0;
+                    if( inPlaneGPU == pmacc::math::Int<simDim>::create(0) )
+                        isGroupRoot = true;
+                }
+
+                algorithm::mpi::Reduce<simDim>* createReduce =
```

Comment:
analogue to #2683 we must add event system barriers before blocking MPI calls `__getTransactionEvent().waitForFinished();`